object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/CalledMethodsPredicateEvaluator.java

```diff
@@ -1,23 +1,181 @@
+
+    try {
+      prover.addConstraint(satQuery);
+    } catch (InterruptedException e) {
+      return false;
+    }
+
+    try {
+      return prover.isUnsat();
+    } catch (Exception e) {
+      return false;
+    }
+  }
+
+  /**
+   * Converts a String representing a boolean expression into a boolean formula in the SMT solver's
+   * format.
+   */
+  private static BooleanFormula formulaStringToBooleanFormula(
+      String formula, BooleanFormulaManager booleanFormulaManager) {
+
+    String classWithFormula = "class DUMMY { boolean DUMMY() { return " + formula + "; } }";
+
+    CompilationUnit ast = parser.parse(classWithFormula).getResult().orElse(null);
+
+    if (ast == null) {
+      throw new BugInCF(
+          "Encountered an unparseable formula while parsing an @CalledMethodsPredicate "
+              + "annotation, but ObjectConstructionVisitor failed to stop compilation. Unparseable formula: "
+              + formula);
+    }
+
+    BlockStmt theBlock = ast.getType(0).getMembers().get(0).asMethodDeclaration().getBody().get();
+
+    com.github.javaparser.ast.expr.Expression theExpression =
+        theBlock.getStatements().get(0).asReturnStmt().getExpression().get();
+
+    return expressionToBooleanFormula(theExpression, booleanFormulaManager);
+  }
+
+  /**
+   * Convert between Javaparser expression and boolean formula via recursive descent.
+   *
+   * @throws UnsupportedOperationException if an unexpected expression is encountered
+   */
+  private static BooleanFormula expressionToBooleanFormula(
+      com.github.javaparser.ast.expr.Expression theExpression,
+      BooleanFormulaManager booleanFormulaManager) {
+    if (theExpression.isNameExpr()) {
+      return booleanFormulaManager.makeVariable(theExpression.asNameExpr().getNameAsString());
+    } else if (theExpression.isBinaryExpr()) {
+      if (theExpression.asBinaryExpr().getOperator().equals(BinaryExpr.Operator.OR)) {
+        return booleanFormulaManager.or(
+            expressionToBooleanFormula(
+                theExpression.asBinaryExpr().getLeft(), booleanFormulaManager),
+            expressionToBooleanFormula(
+                theExpression.asBinaryExpr().getRight(), booleanFormulaManager));
+      } else if (theExpression.asBinaryExpr().getOperator().equals(BinaryExpr.Operator.AND)) {
+        return booleanFormulaManager.and(
+            expressionToBooleanFormula(
+                theExpression.asBinaryExpr().getLeft(), booleanFormulaManager),
+            expressionToBooleanFormula(
+                theExpression.asBinaryExpr().getRight(), booleanFormulaManager));
+      }
+    } else if (theExpression.isEnclosedExpr()) {
+      return expressionToBooleanFormula(
+          theExpression.asEnclosedExpr().getInner(), booleanFormulaManager);
+    } else if (theExpression.isUnaryExpr()) {
+      if (theExpression.asUnaryExpr().getOperator().equals(UnaryExpr.Operator.LOGICAL_COMPLEMENT)) {
```

Comment:
The documentation states that only && and || are supported, not !.

Reply:
The documentation says that `!` is permitted both in the `README` and on the public entry-point that calls this method (`implies`). What documentation are you referring to?

Reply:
Sorry, my mistake.

What I should have said was that the documentation says that all symbols are treated as uninterpreted, but it looks like `false` and `true` might be specially treated.  I'm not sure about that, though.

Reply:
Only `evaluate` treats `true` and `false` as boolean literals. `implies` and its implementation, of which this is part, really do treat all names as uninterpreted. I suspect it will actually fail if a `true` literal is passed, because Javaparser won't turn that into a `NameExpression`, so we'll reach the fall-through case.

Reply:
OK, that sounds good.  I was assuming they treated boolean literals in the same way.